lib/ddtrace/contrib/excon/middleware.rb

```diff
@@ -0,0 +1,140 @@
+require 'ddtrace/propagation/http_propagator'
+
+module Datadog
+  module Contrib
+    module Excon
+      # Middleware implements an excon-middleware for ddtrace instrumentation
+      class Middleware < ::Excon::Middleware::Base
+        SPAN_NAME = 'excon.request'.freeze
+        DEFAULT_ERROR_HANDLER = lambda do |response|
+          Ext::HTTP::ERROR_RANGE.cover?(response[:status])
+        end
+
+        def initialize(stack, options = {})
+          super(stack)
+          @options = Datadog.configuration[:excon].merge(options)
+        end
+
+        def request_call(datum)
+          begin
+            unless datum.key?(:datadog_span)
+              tracer.trace(SPAN_NAME).tap do |span|
+                datum[:datadog_span] = span
+                annotate!(span, datum)
+                propagate!(span, datum) if distributed_tracing?
+              end
+            end
+          rescue StandardError => e
+            Datadog::Tracer.log.debug(e.message)
+          end
+
+          @stack.request_call(datum)
+        end
+
+        def response_call(datum)
+          @stack.response_call(datum).tap do |d|
+            handle_response(d)
+          end
+        end
+
+        def error_call(datum)
+          @stack.error_call(datum).tap do |d|
+            handle_response(d)
+          end
+        end
+
+        # Returns a child class of this trace middleware
+        # With options given as defaults.
+        def self.with(options = {})
+          Class.new(self) do
+            @options = options
+
+            # rubocop:disable Style/TrivialAccessors
+            def self.options
+              @options
+            end
+
+            def initialize(stack)
+              super(stack, self.class.options)
+            end
+          end
+        end
+
+        # Returns a copy of the default stack with the trace middleware injected
+        def self.around_default_stack
+          ::Excon.defaults[:middlewares].dup.tap do |default_stack|
+            # If the default stack contains a version of the trace middleware already...
+            existing_trace_middleware = default_stack.find { |m| m <= Middleware }
+            default_stack.delete(existing_trace_middleware) if existing_trace_middleware
+
+            # Inject after the ResponseParser middleware
+            response_middleware_index = default_stack.index(::Excon::Middleware::ResponseParser).to_i
+            default_stack.insert(response_middleware_index + 1, self)
+          end
+        end
+
+        private
+
+        def tracer
+          @options[:tracer]
+        end
+
+        def distributed_tracing?
+          @options[:distributed_tracing] == true && tracer.enabled
+        end
+
+        def error_handler
+          @options[:error_handler] || DEFAULT_ERROR_HANDLER
+        end
+
+        def split_by_domain?
+          @options[:split_by_domain] == true
+        end
+
+        def annotate!(span, datum)
+          span.resource = datum[:method].to_s.upcase
```

Comment:
The resource in that way is not great. We can keep it that way for now, but we should revisit all our HTTP clients at some point.

Reply:
Agreed. This was how it was before, not sure what it should be. There are a lot of things we could do with this resource.

Reply:
👍